### PR TITLE
Follow-up to h2 storage backend user/password in url: Review feedback switch to regex and move to pure function

### PR DESCRIPTION
**Pull request by oliverse-da**
Follow-up from review feedback of https://github.com/digital-asset/daml/pull/10523

- [x] Switch to using regex
- [x] Unit tests for `extractUserPasswordAndRemoveFromUrl` pure function

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/h2/H2StorageBackend.scala [hunk 1]
```diff
@@ -203,31 +205,29 @@ private[backend] object H2StorageBackend
...
-            )
-        }
-        setter(value)
-        shortenedUrl
+  def extractUserPasswordAndRemoveFromUrl(
+      jdbcUrl: String
+  ): (String, Option[String], Option[String]) = {
+    def setKeyValueAndRemoveFromUrl(url: String, key: String): (String, Option[String]) = {
+      val regex = new Regex(s".*(;${key}=([^;]*)).*")
+      url match {
```
Comment (rautenrieth-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/backend/h2/H2StorageBackend.scala:221]: Don't you need to call `.toLowerCase` somewhere? Or alternatively, use `(?i)` to enable case-insensitive matching?

File: ledger/participant-integration-api/src/test/suite/scala/platform/store/backend/h2/H2StorageBackendSpec.scala [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+package com.daml.platform.store.backend.h2
+
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.wordspec.AnyWordSpec
+
+class H2StorageBackendSpec extends AnyWordSpec with Matchers {
+
+  "H2StorageBackend" should {
+    "extractUserPasswordAndRemoveFromUrl" should {
```
Comment (rautenrieth-da) [inline ledger/participant-integration-api/src/test/suite/scala/platform/store/backend/h2/H2StorageBackendSpec.scala:12]: I would add a test with upper/mixed case `USER`/`PASSWORD` keys. E.g. (from the H2 docs):
```
jdbc:h2:file:~/sample;USER=sa;PASSWORD=123
```

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/h2/H2StorageBackend.scala [hunk 3]
```diff
@@ -203,31 +205,29 @@ private[backend] object H2StorageBackend
...
-              url.take(indexKeyValueBegin) + url.takeRight(url.length - indexKeyValueEnd),
-            )
-        }
-        setter(value)
-        shortenedUrl
+  def extractUserPasswordAndRemoveFromUrl(
+      jdbcUrl: String
+  ): (String, Option[String], Option[String]) = {
+    def setKeyValueAndRemoveFromUrl(url: String, key: String): (String, Option[String]) = {
+      val regex = new Regex(s".*(;${key}=([^;]*)).*")
```
Comment (SamirTalwar-DA) [inline ledger/participant-integration-api/src/main/scala/platform/store/backend/h2/H2StorageBackend.scala:220]: FYI, you can also use the `.r` method on strings. Don't know if you picked this way for readability, but just thought I'd bring it up.

```suggestion
      val regex = s".*(;${key}=([^;]*)).*".r
```

**Comment by oliverse-da** [comment]
Finally getting back to this. Thanks @rautenrieth-da and @SamirTalwar-DA for your feedback. Applied!

**Review by rautenrieth-da — APPROVED** [review]

**Pull request merged**